docs/source/upgrade/index.rst

```diff
@@ -82,13 +83,109 @@ has one *nonconsuming* choice that accepts a *Coin* contract, archives
+
+.. code-block:: none
+
+   $ daml sandbox --port 7001
+
+The version-1 project is named ``coinV1``. It contains just the original *Coin* definition template ``Coin``. Here is the ``daml.yaml`` setup
+
+.. literalinclude:: example/coinV1/daml.yaml
+  :language: yaml
+  :start-after: # BEGIN
+  :end-before: # END
+
+To make the example complete, we need a way to create some coins which have different parties for the issuer and owner. We take the standard approach of adding a :ref:`Propose-Accept workflow <intro propose accept>` to our example.
+
+.. literalinclude:: example/coinV1/daml/CoinV1.daml
+  :language: daml
+  :start-after: -- COIN_TRANSFER_BEGIN
+  :end-before: -- COIN_TRANSFER_END
+
+Now we can build and deploy version-1 of our Coin.
+
+.. code-block:: none
+
+   $ cd example/coinV1
+   $ daml build
+   $ daml deploy --port 7001
+
+
+Create some version-1 coins
+================================
+
+Let's create some coins. We'll use the navigator to connect to the ledger, and create two coins issued by Alice, and owned by Bob.
+
+.. code-block:: none
+
+   $ cd example/coinV1
+   $ daml navigator server localhost 7001
+
+We point a browser to ``localhost:4000``, and follow the steps:
+
+#. Login as Alice:
+    #. Create a coin with Alice as both issuer and owner.
+    #. Exercise the coin's *Transfer* choice, naming Bob as the new owner.
+    #. Create and transfer a 2nd coin in the same way.
+#. Login as Bob:
+    #. Exercise the *Accept* choice on both transfer contracts.
+
+
+Building and deploying version-2
+================================
+
+Now we setup the project for the improved coins containing the *amount* field. The new project is named ``coinV21`` and contains the templates for ``CoinWithAmount``, ``UpgradeCoinProposal`` and ``UpgradeCoinAgreement``. The project setup is:
+
+.. literalinclude:: example/coinV2/daml.yaml
+  :language: yaml
+  :start-after: # BEGIN
+  :end-before: # END
+
+The interesting thing about the project setup for ``coinV2`` is that ``coinV1-0.0.1.dar`` is listed as a dependency. This allows the upgrade templates to make reference to the ``Coin`` template from the ``CoinV1`` module, even though that module is part of original package already deployed to the ledger.
+
+The DAML for the upgrade contacts imports the modules for both the new and old coin versions.
+
+.. literalinclude:: example/coinV2/daml/UpgradeFromCoinV1.daml
+  :language: daml
+  :start-after: -- UPGRADE_MODULE_BEGIN
+  :end-before: -- UPGRADE_MODULE_END
+
+Now we can build and deploy version-2 of our Coin.
+
+.. code-block:: none
+
+   $ cd example/coinV2
+   $ daml build
+   $ daml deploy --port 7001
+
+
+Upgrade existing coins from version-1 to version-2
+================================
+
+We start the navigator again.
+
+.. code-block:: none
+
+   $ cd example/coinV2
+   $ daml navigator server localhost 7001
+
+Finally, we point a browser to ``localhost:4000`` and can effect the coin upgrades:
```

Comment:
Same here, we probably want to change this to a DAML script (in a separate PR).

It would also be nice to explain the trigger-based flow that DAVL uses where you have the following steps:
1. Deploy the upgrade.
2. The issuer runs a script to create `UpgradeProposal`s for all of their coins.
3. The issuer runs a trigger that will upgrade each `Coin` for which there is an `UpgradeCoinAgreement`.
4. The owner interactively (here probably via DAML script (or `daml repl` :slightly_smiling_face:), in a “real” application via some UI) accepts a proposal.

Although it probably makes sense to put this into a separate section and first explain how to do things without a trigger.

Reply:
^ Seconded although doesn't all have to happen in this PR of course - future directions.

Reply:
Yes. This would be great to do.
I've not actually used DAML script myself, so a goo chance to learn.
But we are happy to leave this to a following PR.

Reply:
Having thought a bit about this, I don’t want to use DAML script in those docs for now (at least not as a navigator replacement for the part we currently have, it might still make sense to explain the full DAML script/trigger-based automation used in DAVL). The reason for this is that doing this properly wtr to packaging (which is a crucial part of upgrading so I think this is something that we do need to do properly here) would mean putting the DAML script code into separate packages so we end up with 5 or 6 packages in total which is just too confusing. In principle `daml repl` would be a great navigator replacement for this and avoids the issue (since the code you typo into the repl is never written to a proper package) but `daml repl` is still experimental (for good reasons) and we cannot use an experimental feature in docs for something non-experimental. Therefore, I think for now it makes more sense to add 2 or 3 navigator screenshots.